cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,307 @@
+	entry := LogEntry{
+		Prefix: l.prefix,
+		Level:  l.level,
+		Values: values,
+	}
+	if l.ch != nil {
+		l.ch <- entry
+		return
+	}
+	f, err := Flatten(entry)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(f)
+}
+
+func (l *Logger) clone() *Logger {
+	return &Logger{
+		ch:     l.ch,
+		level:  l.level,
+		prefix: l.prefix,
+		values: copySlice(l.values),
+	}
+}
+
+func copySlice(in []interface{}) []interface{} {
+	out := make([]interface{}, len(in))
+	copy(out, in)
+	return out
+}
+
+// Flatten returns a human readable/machine parsable text representing the LogEntry.
+// Most notable difference with the klog implementation are:
+// - The message is printed at the beginning of the line, without the Msg= variable name e.g.
+//   "Msg"="This is a message" --> This is a message
+// - Variables name are not quoted, eg.
+//   This is a message "Var1"="value" --> This is a message Var1="value"
+// - Variables are not sorted, thus allowing full control to the developer on the output.
+func Flatten(entry LogEntry) (string, error) {
+	var msgValue string
+	var errorValue error
+	if len(entry.Values)%2 == 1 {
+		return "", errors.New("log entry cannot have odd number off keyAndValues")
+	}
+
+	keys := make([]string, 0, len(entry.Values)/2)
+	values := make(map[string]interface{}, len(entry.Values)/2)
+	for i := 0; i < len(entry.Values); i += 2 {
+		k, ok := entry.Values[i].(string)
+		if !ok {
+			panic(fmt.Sprintf("key is not a string: %s", entry.Values[i]))
+		}
+		var v interface{}
+		if i+1 < len(entry.Values) {
+			v = entry.Values[i+1]
+		}
+		switch k {
+		case "msg":
+			msgValue, ok = v.(string)
+			if !ok {
+				panic(fmt.Sprintf("the msg value is not of type string: %s", v))
+			}
+		case "error":
+			errorValue, ok = v.(error)
+			if !ok {
+				panic(fmt.Sprintf("the error value is not of type error: %s", v))
+			}
+		default:
+			if _, ok := values[k]; !ok {
+				keys = append(keys, k)
+			}
+			values[k] = v
+		}
+	}
+	str := ""
+	if entry.Prefix != "" {
+		str += fmt.Sprintf("[%s] ", entry.Prefix)
+	}
+	str += msgValue
+	if errorValue != nil {
+		if msgValue != "" {
+			str += ": "
+		}
+		str += errorValue.Error()
+	}
+	for _, k := range keys {
+		prettyValue, err := pretty(values[k])
+		if err != nil {
+			return "", err
+		}
+		str += fmt.Sprintf(" %s=%s", k, prettyValue)
+	}
+	return str, nil
+}
+
+func pretty(value interface{}) (string, error) {
+	jb, err := json.Marshal(value)
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to marshal %s", value)
+	}
+	return string(jb), nil
+}
+
+// UnstructuredToValues provides a utility function for creating values describing Unstructured objects. e.g.
+// - Deployment="capd-controller-manager" Namespace="capd-system"  (<Kind>=<name> Namespace=<Namespace>)
+// - CustomResourceDefinition="dockerclusters.infrastructure.cluster.x-k8s.io" (omit Namespace if it does not apply)
+func UnstructuredToValues(obj unstructured.Unstructured) []interface{} {
```

Comment:
I don't see this being used anywhere. Is there a use case for this? Maybe it can be describe in a test.

Reply:
It is used in https://github.com/kubernetes-sigs/cluster-api/pull/2191

Reply:
I noticed that PR has a pkg/log/util.go with the same functionality. Maybe we can do without one or the other. But I think it makes more sense to have it in PR #2191 instead of this one.